File: backend/src/domain/entities/report.py
from typing import Any, Dict, List

from pydantic import BaseModel, Field


class DashboardFindingsSummary(BaseModel):
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0
    info: int = 0


class DashboardRiskSummary(BaseModel):
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0


class DashboardSummaryResponse(BaseModel):
    """Platform-wide aggregated metrics."""

    asset_count: int = 0
    internet_exposed_assets: int = 0
    open_ports: int = 0
    services: int = 0
    findings: DashboardFindingsSummary = Field(default_factory=DashboardFindingsSummary)
    risk: DashboardRiskSummary = Field(default_factory=DashboardRiskSummary)


class DashboardTrendPoint(BaseModel):
    """Single point in a time-series trend."""

    timestamp: str
    value: float


class DashboardTrendResponse(BaseModel):
    """Time-series trends for risk and findings."""

    window_days: int = 30
    risk_trend: List[DashboardTrendPoint] = Field(default_factory=list)
    finding_trend: List[DashboardTrendPoint] = Field(default_factory=list)
    critical_finding_trend: List[DashboardTrendPoint] = Field(default_factory=list)


class ExecutiveReportResponse(BaseModel):
    """High-level security posture report."""

    summary: Dict[str, Any] = Field(default_factory=dict)
    top_risky_assets: List[Dict[str, Any]] = Field(default_factory=list)
    critical_findings: List[Dict[str, Any]] = Field(default_factory=list)
    risk_distribution: Dict[str, int] = Field(default_factory=dict)
    exposure_distribution: Dict[str, int] = Field(default_factory=dict)


class AssetReportResponse(BaseModel):
    """Detailed report for a single asset."""

    asset: Dict[str, Any]
    exposure: Dict[str, Any]
    risk: Dict[str, Any]
    ports: List[Dict[str, Any]] = Field(default_factory=list)
    services: List[Dict[str, Any]] = Field(default_factory=list)
    technologies: List[str] = Field(default_factory=list)
    findings: List[Dict[str, Any]] = Field(default_factory=list)


class FindingReportResponse(BaseModel):
    """Filtered list of findings for reporting."""

    findings: List[Dict[str, Any]] = Field(default_factory=list)
    total_count: int = 0


class RiskReportResponse(BaseModel):
    """Detailed risk intelligence report."""

    risk_distribution: Dict[str, int] = Field(default_factory=dict)
    critical_assets: List[Dict[str, Any]] = Field(default_factory=list)
    high_risk_assets: List[Dict[str, Any]] = Field(default_factory=list)
    risk_history: List[Dict[str, Any]] = Field(default_factory=list)


class ExposureReportResponse(BaseModel):
    """Asset distribution by network exposure."""

    external_assets: List[Dict[str, Any]] = Field(default_factory=list)
    internal_assets: List[Dict[str, Any]] = Field(default_factory=list)
    unknown_assets: List[Dict[str, Any]] = Field(default_factory=list)
    internet_exposed_count: int = 0

File: frontend/src/types/report.ts
export interface DashboardFindingsSummary {
  critical: number;
  high: number;
  medium: number;
  low: number;
  info: number;
}

export interface DashboardRiskSummary {
  critical: number;
  high: number;
  medium: number;
  low: number;
}

export interface DashboardSummary {
  asset_count: number;
  internet_exposed_assets: number;
  open_ports: number;
  services: number;
  findings: DashboardFindingsSummary;
  risk: DashboardRiskSummary;
}

export interface DashboardTrendPoint {
  timestamp: string;
  value: number;
}

export interface DashboardTrends {
  window_days: number;
  risk_trend: DashboardTrendPoint[];
  finding_trend: DashboardTrendPoint[];
  critical_finding_trend: DashboardTrendPoint[];
}

export interface ExecutiveReport {
  summary: Record<string, any>;
  top_risky_assets: Record<string, any>[];
  critical_findings: Record<string, any>[];
  risk_distribution: Record<string, number>;
  exposure_distribution: Record<string, number>;
}

export interface AssetReport {
  asset: Record<string, any>;
  exposure: Record<string, any>;
  risk: Record<string, any>;
  ports: Record<string, any>[];
  services: Record<string, any>[];
  technologies: string[];
  findings: Record<string, any>[];
}

export interface FindingReport {
  findings: Record<string, any>[];
  total_count: number;
}

export interface RiskReport {
  risk_distribution: Record<string, number>;
  critical_assets: Record<string, any>[];
  high_risk_assets: Record<string, any>[];
  risk_history: Record<string, any>[];
}

export interface ExposureReport {
  external_assets: Record<string, any>[];
  internal_assets: Record<string, any>[];
  unknown_assets: Record<string, any>[];
  internet_exposed_count: number;
}
